### PR TITLE
HHH-15263 - @NamedQuery is not supported with UPDATE statement

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
https://hibernate.atlassian.net/browse/HHH-15263

Not sure my code changes are totally correct. Feel free to work on top of it or even close it.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java [hunk 1]
```diff
@@ -817,14 +817,18 @@ public LockOptions getLockOptions() {
 
 	@Override
 	public SqmQueryImplementor<R> setLockOptions(LockOptions lockOptions) {
-		verifySelect();
+		if ( lockOptions.getLockMode() != LockMode.NONE ) {
+			verifySelect();
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java:821]:
```suggestion
			// This is required by the JPA spec
			verifySelect();
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java [hunk 2]
```diff
@@ -817,14 +817,18 @@ public LockOptions getLockOptions() {
...
 		getQueryOptions().getLockOptions().overlay( lockOptions );
 		return this;
 	}
 
 	@Override
 	public SqmQueryImplementor<R> setLockMode(String alias, LockMode lockMode) {
-		verifySelect();
+		if ( lockMode != LockMode.NONE ) {
+			verifySelect();
+		}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java:831]: IMO we should remove this restriction, as that is a Hibernate method and we actually do support locking in subqueries. Wdyt @dreab8, @sebersole ?

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java:831]: I'm fine with removing this check here

**Comment by NathanQingyangXu** [comment]
Steve, is it because JPA has the feature that only select query can lock at
the same time?

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java (hunk 1 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java:821]: I can add it, but there is similar comment within `verifySelect()` already:
```
	protected void verifySelect() {
		try {
			SqmUtil.verifyIsSelectStatement( getSqmStatement(), hql );
		}
		catch (IllegalQueryOperationException e) {
			// per JPA
			throw new IllegalStateException( "Expecting a SELECT query : `" + hql + "`", e );
		}
	}
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java (hunk 2 above)
Comment (NathanQingyangXu) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java:831]: I made the changes to be in line with another method in the same class:
```
	@Override
	public SqmQueryImplementor<R> setLockMode(LockModeType lockMode) {
		if ( lockMode != LockModeType.NONE ) {
			verifySelect();
		}
		getSession().checkOpen( false );
		getQueryOptions().getLockOptions().setLockMode( LockMode.fromJpaLockMode( lockMode ) );
		return this;
	}
```
@sebersole , saw the code was written recently. Should we remove this check (maybe everywhere in this class)?

**Comment by beikov** [comment]
Thanks

**Pull request merged**